src/main/java/ca/on/oicr/pineryreports/util/SampleUtils.java

```diff
@@ -0,0 +1,225 @@
+    return getIntAttribute(attributeName, sample, null);
+  }
+  
+  /**
+   * Return attribute value that exists for the given attribute name on the given sample.
+   * 
+   * @param sample Sample to look for attribute
+   * @param attributeName Name of the attribute we are retrieving values for
+   * @param defaultValue value to return if the attribute is not found
+   * @return Attribute the attribute value, or defaultValue if the attribute is not found
+   */
+  public static Integer getIntAttribute(String attributeName, SampleDto sample, Integer defaultValue) {
+    String attr = getAttribute(attributeName, sample);
+    return attr == null ? defaultValue : Integer.valueOf(attr);
+  }
+  
+  /**
+   * Get an attribute from higher up in the hierarchy. Will never return an attribute found directly on sample
+   * 
+   * @param attributeName
+   * @param sample find the attribute in this sample's hierarchy
+   * @param allSamples complete set of potential ancestors to this sample, mapped by ID
+   * @return the attribute value from the nearest ancestor which has the attribute, or null if not found
+   */
+  public static String getUpstreamAttribute(String attributeName, SampleDto sample, Map<String, SampleDto> allSamples) {
+    for (SampleDto parent = getParent(sample, allSamples); parent != null; parent = getParent(parent, allSamples)) {
+      String value = getAttribute(attributeName, parent);
+      if (value != null) return value;
+    }
+    return null;
+  }
+  
+  /**
+   * Get the direct parent of sample
+   * 
+   * @param sample sample to find the parent of
+   * @param potentialParents complete set of potential ancestors to this sample, mapped by ID
+   * @return the sample's parent, or null if it has no parent
+   */
+  public static SampleDto getParent(SampleDto sample, Map<String, SampleDto> potentialParents) {
+    String parentId = getParentId(sample);
+    if (parentId == null) {
+      return null;
+    }
+    SampleDto parent = potentialParents.get(parentId);
+    if (parent == null) {
+      throw new IllegalStateException("Parent sample " + parentId + " not found. Possibly in a different project");
+    }
+    return parent;
+  }
+  
+  /**
+   * Get an ancestor of the sample
+   * 
+   * @param sample sample to find the parent of
+   * @param sampleCategory category of parent to find
+   * @param potentialParents complete set of potential ancestors to this sample, mapped by ID
+   * @return the closest ancestor of sample of the provided sample category
+   */
+  public static SampleDto getParent(SampleDto sample, String sampleCategory, Map<String, SampleDto> potentialParents) {
+    for (SampleDto current = sample; current != null; current = getParent(current, potentialParents)) {
+      if (sampleCategory.equals(getAttribute("Sample Category", current))) {
+        return current;
+      }
+    }
+    throw new IllegalStateException("Parent " + sampleCategory + " of sample " + sample.getId()
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "sample" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)